src/ConnectionPool.jl

```diff
@@ -0,0 +1,575 @@
+        wait(t.c.readdone)
+    end                                           ;@debug 2 "👁  Start read: $t"
+    t.c.readbusy = true
+    @ensure isreadable(t)
+    return
+end
+
+
+"""
+    closeread(::Transaction)
+
+Signal that an entire Response Message has been read from the `Transaction`.
+
+Increment `readcount` and wake up tasks waiting in `startread`.
+"""
+
+function IOExtras.closeread(t::Transaction)
+    @require isreadable(t)
+
+    t.c.readbusy = false
+    t.c.readcount += 1
+    notify(t.c.readdone)                          ;@debug 2 "✉️  Read done:  $t"
+    notify(poolcondition)
+
+    @ensure !isreadable(t)
+    return
+end
+
+function Base.close(t::Transaction)
+    close(t.c)
+    if iswritable(t)
+        closewrite(t)
+    end
+    if isreadable(t)
+        closeread(t)
+    end
+    return
+end
+
+function Base.close(c::Connection)
+    close(c.io)
+    if nb_available(c) > 0
+        purge(c)
+    end
+    notify(poolcondition)
+    return
+end
+
+
+"""
+    purge(::Connection)
+
+Remove unread data from a `Connection`.
+"""
+
+function purge(c::Connection)
+    @require !isopen(c.io)
+    while !eof(c.io)
+        readavailable(c.io)
+    end
+    c.excess = nobytes
+    @ensure nb_available(c) == 0
+end
+
+
+"""
+The `pool` is a collection of open `Connection`s.  The `request`
+function calls `getconnection` to retrieve a connection from the
+`pool`.  When the `request` function has written a Request Message
+it calls `closewrite` to signal that the `Connection` can be reused
+for writing (to send the next Request). When the `request` function
+has read the Response Message it calls `closeread` to signal that
+the `Connection` can be reused for reading.
+"""
+
+const pool = Vector{Connection}()
+const poollock = ReentrantLock()
+const poolcondition = Condition()
+
+"""
+    closeall()
+
+Close all connections in `pool`.
+"""
+
+function closeall()
+
+    lock(poollock)
+    for c in pool
+        close(c)
+    end
+    empty!(pool)
+    unlock(poollock)
+    notify(poolcondition)
+    return
+end
+
+
+"""
+    findwritable(type, host, port) -> Vector{Connection}
+
+Find `Connections` in the `pool` that are ready for writing.
+"""
+
+function findwritable(T::Type,
+                      host::AbstractString,
+                      port::AbstractString,
+                      pipeline_limit::Int,
+                      reuse_limit::Int)
+
+    filter(c->(!c.writebusy &&
+               typeof(c.io) == T &&
+               c.host == host &&
+               c.port == port &&
+               c.pipeline_limit == pipeline_limit &&
+               c.writecount < reuse_limit &&
+               c.writecount - c.readcount < pipeline_limit + 1 &&
+               isopen(c.io)), pool)
+end
+
+
+"""
+    findoverused(type, host, port, reuse_limit) -> Vector{Connection}
+
+Find `Connections` in the `pool` that are over the reuse limit
+and have no more active readers.
+"""
+
+function findoverused(T::Type,
```

Comment:
I'd be interested to hear where this notion of `overuse` comes from; just general good practice? some RFC somewhere? I ask because I haven't heard of only re-using a connection so many times.

Reply:
Most services seem to have a limit on connection reuse. e.g. httpbin seems to disconnect after exactly 100 requests. This post suggests that S3 has a similar limit: https://forums.aws.amazon.com/thread.jspa?threadID=91402, but the doc referred to no longer specifies this. My guess is that S3 now has a dynamic limit on the number of requests per connection. In my own testing it sometimes seems to drop at about 100, but sometimes as low as 40. The issue is that if you send out 120 pipelined requests, and the server disconnects when it gets up to reading the 90th request, you have to re-transmit 30 requests. So it's best to figure out a limit safely lower than the servers you're dealing with. Hence `findoverused`. Without this feature, the performance of the massively parallel S3 and httpbin tests was abysmal. My guess is that the server probably throttles/penalises clients who repeatedly pipeline more requests than the server will accept.